AsyncDisplayKit/Details/ASCollectionViewFlowLayoutInspector.m

```diff
@@ -0,0 +1,121 @@
+ * of patent rights can be found in the PATENTS file in the same directory.
+ */
+
+#import <UIKit/UIKit.h>
+
+#import "ASCollectionViewFlowLayoutInspector.h"
+
+#import "ASCollectionView.h"
+
+@implementation ASCollectionViewFlowLayoutInspector {
+  BOOL _delegateImplementsReferenceSizeForHeader;
+  BOOL _delegateImplementsReferenceSizeForFooter;
+}
+
+#pragma mark - Accessors
+
+- (instancetype)initWithCollectionView:(ASCollectionView *)collectionView flowLayout:(UICollectionViewFlowLayout *)flowLayout;
+{
+  self = [super init];
+
+  if (flowLayout == nil) {
+    return nil;
+  }
+
+  if (self != nil) {
+    [self cacheSelectorsForCollectionView:collectionView];
+    _layout = flowLayout;
+  }
+  return self;
+}
+
+- (void)cacheSelectorsForCollectionView:(ASCollectionView *)collectionView
+{
+  if (collectionView == nil) {
+    _delegateImplementsReferenceSizeForHeader = nil;
+    _delegateImplementsReferenceSizeForFooter = nil;
+  } else {
+    _delegateImplementsReferenceSizeForHeader = [[self delegateForCollectionView:collectionView] respondsToSelector:@selector(collectionView:layout:referenceSizeForHeaderInSection:)];
+    _delegateImplementsReferenceSizeForFooter = [[self delegateForCollectionView:collectionView] respondsToSelector:@selector(collectionView:layout:referenceSizeForFooterInSection:)];
+  }
+}
+
+#pragma mark - ASCollectionViewLayoutInspecting
+
+- (ASSizeRange)collectionView:(ASCollectionView *)collectionView constrainedSizeForSupplementaryNodeOfKind:(NSString *)kind atIndexPath:(NSIndexPath *)indexPath
+{
+  CGSize constrainedSize;
+  CGSize supplementarySize = [self sizeForSupplementaryViewOfKind:kind inSection:indexPath.section collectionView:collectionView];
+  if (_layout.scrollDirection == UICollectionViewScrollDirectionVertical) {
+    constrainedSize = CGSizeMake(collectionView.bounds.size.width, supplementarySize.height);
+  } else {
+    constrainedSize = CGSizeMake(supplementarySize.height, collectionView.bounds.size.height);
+  }
+  return ASSizeRangeMake(CGSizeZero, constrainedSize);
+}
+
+- (NSUInteger)collectionView:(ASCollectionView *)collectionView numberOfSectionsForSupplementaryNodeOfKind:(NSString *)kind
+{
+  if ([collectionView.asyncDataSource respondsToSelector:@selector(numberOfSectionsInCollectionView:)]) {
+    return [collectionView.asyncDataSource numberOfSectionsInCollectionView:collectionView];
+  } else {
+    return 1;
+  }
+}
+
+- (NSUInteger)collectionView:(ASCollectionView *)collectionView supplementaryNodesOfKind:(NSString *)kind inSection:(NSUInteger)section
+{
+  return [self layoutHasSupplementaryViewOfKind:kind inSection:section collectionView:collectionView] ? 1 : 0;
+}
+
+#pragma mark - Private helpers
+
+- (CGSize)sizeForSupplementaryViewOfKind:(NSString *)kind inSection:(NSUInteger)section collectionView:(ASCollectionView *)collectionView
+{
+  if ([kind isEqualToString:UICollectionElementKindSectionHeader]) {
```

Comment:
Would using isEqual: offer a little performance bump by trying an identity comparison before a deep compare? IIRC that's how the isEqual:/isEqualToBanana: methods work.

Sent from my iPhone

> On Oct 18, 2015, at 11:07 PM, appleguy <notifications@github.com> wrote:
> 
> In AsyncDisplayKit/Details/ASCollectionViewFlowLayoutInspector.m:
> 
> > +    return [collectionView.asyncDataSource numberOfSectionsInCollectionView:collectionView];
> > +  } else {
> > +    return 1;
> > +  }
> > +}
> > +
> > +- (NSUInteger)collectionView:(ASCollectionView *)collectionView supplementaryNodesOfKind:(NSString *)kind inSection:(NSUInteger)section
> > +{
> > +  return [self layoutHasSupplementaryViewOfKind:kind inSection:section collectionView:collectionView] ? 1 : 0;
> > +}
> > +
> > +#pragma mark - Private helpers
> > +
> > +- (CGSize)sizeForSupplementaryViewOfKind:(NSString *)kind inSection:(NSUInteger)section collectionView:(ASCollectionView *)collectionView
> > +{
> > +  if ([kind isEqualToString:UICollectionElementKindSectionHeader]) {
> ak, kinda kills me to see an isEqualToString: in a method like this :). I wish Apple designed this a bit better, even an enum would be nice!
> 
> I think the implementation of this method is the most notable and non-ASDK approach in the patch, and we should make sure to support cell/supplementary-specified sizing soon.
> 
> —
> Reply to this email directly or view it on GitHub.
>